### PR TITLE
rustc_target: `TyAndLayout::field` should never error.

### PR DESCRIPTION
**Pull request by eddyb**
This refactor (making `TyAndLayout::field` return `TyAndLayout` without any `Result` around it) is based on a simple observation, regarding `TyAndLayout::field`:

If `cx.layout_of(ty)` succeeds (for some `cx` and `ty`), then `.field(cx, i)` on the resulting `TyAndLayout` should *always* succeed in computing `cx.layout_of(field_ty)` (where `field_ty` is the type of the `i`th field of `ty`).

The reason for this is that no matter which field is chosen, `cx.layout_of(field_ty)` *will have already been computed*, as part of computing `cx.layout_of(ty)`, as we cannot determine the layout of *any* type without considering the layouts of *all* of its fields.

And so it should be fine to turn any errors into ICEs, since they likely indicate a `cx` mismatch, or some other edge case that is due to a compiler bug (as opposed to ever being an user-facing error).

<hr/>

Each commit should probably be reviewed separately, though note that there's some `where` clauses (in `rustc_target::abi::call::*`) that change in most commits.

cc @nagisa @oli-obk

### CODE REVIEW

File: compiler/rustc_codegen_ssa/src/mir/block.rs [hunk 1]
```diff
@@ -473,9 +473,9 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
             let do_panic = match intrinsic {
                 Inhabited => layout.abi.is_uninhabited(),
                 // We unwrap as the error type is `!`.
-                ZeroValid => !layout.might_permit_raw_init(bx, /*zero:*/ true).unwrap(),
+                ZeroValid => !layout.might_permit_raw_init(bx, /*zero:*/ true),
                 // We unwrap as the error type is `!`.
```
Comment (nagisa) [inline compiler/rustc_codegen_ssa/src/mir/block.rs:477]: The comment here is no longer relevant.

File: compiler/rustc_middle/src/ty/layout.rs [hunk 2]
```diff
@@ -2134,32 +2144,24 @@ where
         TyAndLayout { ty: this.ty, layout }
     }
 
-    fn field(this: TyAndLayout<'tcx>, cx: &C, i: usize) -> C::TyAndLayout {
-        enum TyMaybeWithLayout<C: LayoutOf> {
-            Ty(C::Ty),
-            TyAndLayout(C::TyAndLayout),
+    fn ty_and_layout_field(this: TyAndLayout<'tcx>, cx: &C, i: usize) -> TyAndLayout<'tcx> {
```
Comment (nagisa) [inline compiler/rustc_middle/src/ty/layout.rs:2147]: Hm, would there be value in making sure the naming between this and the `ty_and_layout_FOR_variant` is consistent?

**Comment by nagisa** [comment]
r=me up to you to decide if you want to do anything about the inline comments.

File: compiler/rustc_middle/src/ty/layout.rs (hunk 2 above)
Comment (eddyb) [reply compiler/rustc_middle/src/ty/layout.rs:2147]: The prefix `ty_and_layout_` is consistent between the two, IIRC. If you mean `TyAndLayout::for_variant` should be renamed, I guess `as_variant` or `variant_downcast` might be okay, but I don't have a strong preference.

Comment (nagisa) [reply compiler/rustc_middle/src/ty/layout.rs:2147]: Yeah I mean `ty_and_layout_for_variant`+`ty_and_layout_for_field` or just `ty_and_layout_variant`+`ty_and_layout_field`.

**Comment by eddyb** [comment]
@bors r=nagisa

I resolved the first comment, but my reasoning for not changing the `ty_and_layout_{for_variant,field}` methods that provide `TyAndLayout::{for_variant,field}` (as stated in https://github.com/rust-lang/rust/pull/88337#discussion_r697907576) is that they're consistent right now (via `TyAndLayout::` <-> `ty_and_layout_`).

**Pull request merged**